support dynamic function calls in component model

dicej opened the pull request:
This addresses #4310, introducing a new `component::values::Val` type for
representing component values dynamically.  It also adds a `call` method to
`component::func::Func`, which takes a slice of `Val`s as parameters and returns
a `Result<Val>` representing the result.

As an implementation detail, I've also added a `component::values::Type` type,
which is an owned, despecialized version of
`wasmtime_environ::component::InterfaceType`.  That serves two purposes:

- It allows us to despecialize as the first step, which reduces the number of cases to consider when typechecking and lowering.
- It avoids needing to borrow the store both mutably and immutably when lowering, as we would if we used `InterfaceType`s.

Finally, I happened to notice that the `ComponentType::SIZE32` calculation in
`expand_record_for_component_type` needed a correction, so I did that.

Signed-off-by: Joel Dice <joel.dice@fermyon.com>

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

dicej commented:
@alexcrichton @jameysharp This addresses #4310.  I plan to push one more commit to add "sad path" tests, but this should be complete otherwise.

Please pay particular attention to the `unsafe` block in `Func::call` -- I copied it from `TypedFunc::call_raw` and modified it, but I want to make sure I did that correctly.

dicej commented:
I could use some advice about how to address the CI issue.  I've added a new `wasmtime-component-util` crate with code that I wanted to share between `wasmtime` and `wasmtime-component-macro`.  The `publish` script doesn't like that, though:

```
Caused by:
  no matching package named `wasmtime-component-util` found
  location searched: registry `crates-io`
  required by package `wasmtime-component-macro v0.40.0 (/home/runner/work/wasmtime/wasmtime/target/package/wasmtime-component-macro-0.40.0)`
thread 'main' panicked at 'failed to verify "crates/component-macro/Cargo.toml"', scripts/publish.rs:462:9
```

Is there a better way to reuse that code, or else some way to bootstrap the dependency and make `publish` happy?

jameysharp commented:
I didn't know this, but there's a comment in `publish.rs` that says "note that this list must be topologically sorted by dependencies". Does it work if you swap the order of `wasmtime-component-util` and `wasmtime-component-macro`?

dicej commented:
> I didn't know this, but there's a comment in `publish.rs` that says "note that this list must be topologically sorted by dependencies". Does it work if you swap the order of `wasmtime-component-util` and `wasmtime-component-macro`?

Oh good call.  I _did_ know that, but forgot about it.  Will try that now.

jameysharp commented on crates/wasmtime/src/component/func.rs:
```diff
@@ -241,4 +242,155 @@ impl Func {
+                params
+                    .iter()
+                    .zip(args)
+                    .map(|(ty, arg)| arg.lower(store, &options, ty, &mut space))
+                    .collect::<Result<()>>()
```
I think you can use `Iterator::try_for_each` here for a little more clarity.

jameysharp reviewed (commented):
I'll leave it to @alexcrichton to decide whether to approve this, but I have a few small observations scattered through.

Overall I'm happy with this PR. I'm disappointed at how much `crates/wasmtime/src/component/values.rs` duplicates code that's in `crates/wasmtime/src/component/func/typed.rs` and elsewhere, but I don't see a solid alternative right now, so let's run with it I guess.

@alexcrichton, I think you can focus your review on the changes in `crates/wasmtime/src/component/func.rs`. I feel pretty confident approving the rest, but that file has the one new `unsafe` block and generally the trickiest interactions with other parts of wasmtime.

dicej commented on crates/wasmtime/src/component/func/typed.rs:
```diff
@@ -1095,7 +1093,7 @@ impl WasmStr {
     }
 
-    fn _to_str<'a>(&self, store: &'a StoreOpaque) -> Result<Cow<'a, str>> {
+    pub(crate) fn _to_str<'a>(&self, store: &'a StoreOpaque) -> Result<Cow<'a, str>> {
```
Yeah, that felt awkward to me, too.  Perhaps we could rename it to something like `to_str_with_store_opaque`?  I haven't yet wrapped my head around all the different flavors of store, so perhaps there's a better option.

alexcrichton reviewed (commented):
Thanks for this! Personally I think that this is a really tricky problem and I don't think that there's any good answer right now to how to implement this. On one hand I think that this PR is probably sufficient from and implementation and functionality point of view, but on the other hand I don't think that this is suitable for any performance-conscious scenario. I don't know if this is the best choice of tradeoffs at this time because to me at least we're generally trying to get to a semi-final state for the component model where the known deficiencies are accounted for.

I'm personally wary of throwing in the towel on performance so soon here but I also don't have any alternatives. Even if we go with a strategy like the one you've implemented here there's a lot of open questions to me which I don't think have a clear answer:

* One thing is how types are represented. I would naively expect to be able to, given a `Val`, ask for its type along the lines of the [`wasmtime::Val::ty` method](https://docs.rs/wasmtime/latest/wasmtime/enum.Val.html#method.ty). Given this current implementation I don't think that this is possible. I think, though, it might be possible to add this in where a "type index" of some sort stood next to the `Record`/`Variant`/`Flags` types.

* On the topic of type information one thing that we'll eventually want to add to the component model implementation is the ability to introspect a component for the types of its imports and exports. For example currently Wasmtime has [`wasmtime::Module::imports`](https://docs.rs/wasmtime/latest/wasmtime/struct.Module.html#method.imports) and [`wasmtime::Module::exports`](https://docs.rs/wasmtime/latest/wasmtime/struct.Module.html#method.exports) but the equivalent for the component model isn't implemented yet. I believe an implementation of this will require something along the lines of the `enum Type` you've started here but I don't think we want an "owned `Type`" as you've sketched out which recursively owns all of its contents. Instead I think we'll want a design similar to the `ComponentTypesBuilder` interface right now where there's a "bag of everything" and types all index into that. I don't think that you need to implement `Component::{imports,exports}` in this PR here, but at the same time I'd prefer to avoid having another internal representation of types different from what we already have. Adding this type information may be inevitable for this API because I do think that records/variants at least should be created with their type information up-front (rather than "if you have the right number of fields it all works out")

* One thing I've worried about historically is the size/alignment calculation. Right now in the typed API it's generally all statically done at compile time but with this dynamic API I'm a little worried that with a recursively-defined size calculation it runs the risk of being $O(n^2)$ performance if the size is requested at many layers of the type tree. It might be better to start storing canonical size/align information for aggregates in the `ComponentTypes` structure rather than calculating it on the fly here.

That's at least an initial dump of thoughts I have on the internals here. I wish I had a good answer for this like this ahead of time. Otherwise though while I don't think we need to get this PR to a perfect spot before landing I do think we should get it into a state where the major pieces of functionality are there. For me those are:

* There shouldn't be a duplicate internal representation of type information (currently `enum Type` is crate-local)
* Personally I think that `Val::type(maybe_some_arg) -> Something` should exist  to get the type from a standalone `Val`

Those two points may have a significant degree of fallout depending on the implementation direction.

---

On the topic of code duplication, to some degree that's just life in the component model. I've written the same size/alignment calculation in maybe 4 crates now at this point. It's slightly different each time and I've found it really hard to share code. In general I don't think my development strategy here has been really all that tenable or scalable historically and I'd ideally like to change it.

Where possible I think thinks should be shared. For example reusing `Lift::load` here and `Lower::store` I think is a great idea. If possible it might be good to try and reuse `ComponentType::typecheck` here as well. Otherwise though if something is difficult to share I don't think it's worth bending over backwards to have precisely one implementation of something. To that end that's also why we want a strong fuzzing implementation here eventually.

Specifically for `Func::call` though I would ideally like to share more of that with `TypedFunc::call`. There's a fair amount of nontrivial logic, especially around handling `post_return`, `may_enter`, etc. If you would be ok with it I think it would be good to try to make a generic helper so that `call_unchecked_raw` is only called once or something roughly along those lines. Again though I wouldn't bend over backwards too much for this, I just think it's good to try to investigate and see how it works out.

---

Otherwise as a random final note I think that you'll want to move `post_return` from `TypedFunc` to `Func`. Additionally I think you'll want to assert specific errors on function calls since I think all the `.is_err()` assertions are passing due to "needs a call to `post_return`" rather than type-checking errors that should otherwise be happening.

dicej commented:
Thanks @alexcrichton and @jameysharp for your thoughtful and detailed feedback.

Alex: I agree with what you said about representing types and wanting to be able to get a type from a `Val`.  For my first pass at this, I didn't have a `Type` enum -- I just used `InterfaceType` everywhere.  That worked pretty well until the borrow checker made its verdict.  I couldn't figure out how to pass a `&ComponentTypes` to `Val::lower` since it also takes a `&mut StoreContextMut<T>` since the former was borrowed from the latter.  Now that I reflect on it, though -- I see I don't need to pass both of those things since I can always get the former from the latter "just in time" by reborrowing when needed.  I'll give that another shot and see what happens.  I'll also see what I can do about deduplicating  `Func::call` and `TypedFunc::call`.

alexcrichton commented:
Hm actually I do think you'll want to pass around `&ComponentTypes` if you can. One trick I used with instances is that I took it out of the store temporarily, placing it back into the store when I was done with it during the function call.

I think this will want to be an argument because if an end user creates a `Val` they'll need to create it with a type for cases like a `Record`, so the end-user may create a foreign `ComponentTypes`. Hm maybe at least, I'm not sure. They'll always want to use types already registered within the component so perhaps not... Well in any case as long as it works!

dicej commented:
I've tried a few variations.  Here's the latest:

```rust
pub struct Types<'a>(&'a ComponentTypes);

pub struct Case<'a> {
    name: &'a str,
    ty: Type,
}

#[derive(Copy, Clone)]
pub struct VariantIndex(TypeVariantIndex);

impl VariantIndex {
    pub fn name(self, types: Types) -> &str {
        &types.0[self.0].name
    }

    pub fn cases(self, types: Types) -> impl ExactSizeIterator<Case> {
        types.0[self.0].cases.iter().map(|case| Case {
            name: &case.name,
            ty: Type::from(&case.ty),
        })
    }
}

pub enum Type {
    // ...
    Variant(VariantIndex),
    // ...
}

impl Type {
    fn from(ty: &InterfaceType) -> Self {
        match ty {
            // ...
            InterfaceType::Variant(index) => Type::Variant(VariantIndex(*index)),
            // ...
        }
    }
}

pub enum Val {
    // ...
    Variant(Variant),
    // ...
}

impl Val {
    pub fn ty(&self) -> Type {
        match self {
            // ...
            Self::Variant { ty, .. } => Type::Variant(ty),
            // ...
        }
    }

    // ...
}

pub struct Variant {
    ty: VariantIndex,
    discriminant: u32,
    value: Box<Val>,
}

impl Variant {
    pub fn try_new(
        ty: VariantIndex,
        types: Types,
        discriminant: u32,
        value: Box<Val>,
    ) -> Result<Self> {
        // ... (here we would verify that `discriminant` and `value` match `ty` and throw an error otherwise)

        Ok(Self {
            ty,
            discriminant,
            value,
        })
    }
}

impl Func {
    pub fn types(&self, store: impl AsContext<'a>) -> Types<'a> {
        Types(&store.as_context()[self.0].types)
    }

    pub fn get_param_type(&self, store: impl AsContext, index: u32) -> Type {
        let data = &store.as_context()[self.0];
        Type::from(data.types[data.ty].params[index])
    }

    pub fn get_result_type(&self, store: impl AsContext) -> Type {
        let data = &store.as_context()[self.0];
        Type::from(data.types[data.ty].result)
    }

    // ...
}

#[test]
fn foo() -> Result<()> {
    let engine = super::engine();
    let mut store = Store::new(&engine, ());
    let component = Component::new(
        &engine,
        make_echo_component_with_params(
            r#"(variant (case "A" u32) (case "B" s32))"#,
            &[super::Type::I32, super::Type::I32],
        ),
    )?;
    let instance = Linker::new(&engine).instantiate(&mut store, &component)?;
    let func = instance.get_func(&mut store, "echo").unwrap();
    let ty = func.get_param_type(&store, 0);

    if let Type::Variant(index) = ty {
        let input = [Val::Variant(Variant::try_new(ty, func.types(&store))?)];
        let output = func.call(&mut store, &input)?;

        assert_eq!(input[0], output)
    } else {
        unreachable!()
    }

    Ok(())
}
```

The ergonomics aren't outstanding, but the efficiency is pretty good.  My rationale for still exposing indexes is that Wasm itself uses indexes everywhere, so the paradigm will at least be familiar to users.

Happy to just use an `Arc<ComponentTypes>` everywhere if we want to prioritize ergonomics, though.

alexcrichton commented:
Apologies for the delay, busy day yesterday and this morning! That looks good to me though. Some things may be tweaked over time but I think that's good to land.

dicej commented:
I've overhauled the code according to the above feedback, and I decided to squash and rebase since so much was changing anyway.  Please let me know what you think.

Highlights to note:
- `component::types::Type` is public, and `Val` now has a `ty` method to get the type of that value as a `Type`.  This should also be suitable for use with `Component::exports` and `Component::imports` when the time comes to implement those.
- I manged to reuse `call_raw` so there's minimal duplication in `Func::call` and no new `unsafe` code (but note that `call_raw` has moved to `Func` along with `post_return`).
- Type checking now mostly happens during `Val` creation, with one final, quick, linear-time check happening in `Func::call`.
- The sad path tests are now checking for specific errors.

alexcrichton commented on crates/wasmtime/src/component/func/typed.rs:
```diff
@@ -730,6 +510,27 @@ forward_impls! {
+    std::rc::Rc<str>,
+    std::sync::Arc<str>,
+    String,
+}
```
While you're here and doing this could this also add variants for lifting `Vec<T>` from `WasmSlice<T>` and such?

The macro could possibly look something like:

```rust
forward_lifts! {
    WasmStr => { Box<str>, Rc<str>, Arc<str>, String }
    WasmList<T> => { Box<[T]>, ...}
}
```

(or similar)

alexcrichton commented on crates/wasmtime/src/component/func.rs:
```diff
@@ -241,4 +249,342 @@ impl Func {
+                    self.store_args(store, &options, &params, args, dst)
+                } else {
+                    args.iter()
+                        .try_for_each(|arg| arg.lower(store, &options, dst, 0))
```
This seems like it may not be right for multiple arguments since the offset may always be zero?

alexcrichton commented on crates/wasmtime/src/component/func.rs:
```diff
@@ -241,4 +249,342 @@ impl Func {
+    fn load_result<'a>(
+        mem: &Memory,
+        ty: &Type,
+        src: &mut impl Iterator<Item = &'a ValRaw>,
```
I worry with `impl Iterator` that this will excessively monomorphize into callers, so could this use something concrete like `&mut &[ValRaw]` or `&mut std::slice::Iter<'_, ValRaw>`?

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(if *value { 1 } else { 0 }));
+            }
+            Val::S8(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value as i32));
```
Idiomatically we try to avoid `as` casts where possible since in isolation they could indicate that a lossy conversion is happening. Could the casts here be replaced with `i32::from(*value)`? That should signify that no data is lost.

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+            }
+            Val::Record(Record { values, .. }) | Val::Tuple(Tuple { values, .. }) => {
+                for (index, value) in values.iter().enumerate() {
+                    value.lower(store, options, dst, offset + index)?;
```
I think this `+ index` may not be right since each of the `values` may take up more than one `ValRaw` slot.

It might make more sense for this function to take `&mut std::slice::Iter<'_, MaybeUninit<ValRaw>>` instead of the slice/offset it currently has perhaps?

alexcrichton replied:
One nice part of perhaps using `MaybeUninit<ValRaw>` is that you could delegate to the `Lower for T` implementations? Or otherwise I think it would be worthwhile figuring out how to delegate to that for primitives and strings/lists perhaps (may require an unsafe method or two to extract `MaybeUninit<[ValRaw; 2]>` from `&mut std::slice::Iter<'_, MaybeUninit<ValRaw>>` though.

dicej replied:
@alexcrichton Is there a convenient way to make a `&mut std::slice::Iter<'_, &mut MaybeUninit<ValRaw>>` from a `&mut MaybeUninit<[ValRaw; N]`?  I could do something like `(0..N).map(|i| map_maybe_uninit!(dst[i])`, but I don't think that's what you meant.

alexcrichton replied:
Not easily, but you can transmute `&mut MaybeUninit<[ValRaw; N]>` to `&mut [MaybeUninit<ValRaw>; N]` "safely" and then `.iter()` from there

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+                ..
+            }) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*discriminant));
+                value.lower(store, options, dst, offset + 1)?;
```
This is another case that I think is not quite right for subtle reasons. The main one is that unused `ValRaw` entries (and upper bits which are unused here) need to all be set to zero. Otherwise this runs the risk of leaking undefined values into wasm modules.

Additionally this is another location that needs to inform the caller that a nonzero amount of `ValRaw` slots are consumed. Given `value` here you may also consume more slots than `value` takes up because of how variants are "flattened"

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+
+    /// Serialize this value to the heap at the specified memory location.
+    pub(crate) fn store<T>(&self, mem: &mut MemoryMut<'_, T>, offset: usize) -> Result<()> {
+        match self {
```
Could this have a preceding `debug_assert!` about alignment?

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+                // FIXME: needs memory64 handling
+                *mem.get(offset + 0) = (ptr as i32).to_le_bytes();
+                *mem.get(offset + 4) = (len as i32).to_le_bytes();
+            }
```
I think this could delegate to `<str as Lower>::store`?

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,389 @@
+fn lower_list<T>(mem: &mut MemoryMut<'_, T>, items: &[Val]) -> Result<(usize, usize)> {
+    let element_type = match items {
+        [item, ..] => item.ty(),
+        [] => Type::Unit,
```
I think this is actually subtly incorrect because it would get the alignment incorrect for the call to `realloc`. The type of the list element may need to get passed in here.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+
+impl<T: PartialEq> PartialEq for Handle<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index && Arc::ptr_eq(&self.types, &other.types)
```
I think that this may lead to unfortunately surprising results in that a string type is only considered equivalent if it comes from the same `types` origin but it's actually unique across all components. I think that this is sufficient for a "succeed fast" test but this may need to be expanded for deep equality checking as well.

dicej replied:
I'm not sure I understand this.  `Type::String` doesn't have a `Handle<T>`, so this code wouldn't be reached in that case.  Am I misunderstanding what you meant by "string type"?

alexcrichton replied:
Ah sorry bad example, but instead I think this would consider `(u32, u32)` not an equivalent type if it comes from two different components.

dicej replied:
I've address this for tuple, option, and expected.  Not sure if I should do the same for e.g. record, though.  See my comment in the other conversation about type equality.

alexcrichton replied:
I think that the tuple/option/expected bits are probably not going to be necessary in the long-term if the original issue here is fixed. I think this is fine to leave as a FIXME with a follow-up issue for now though.

Otherwise though the current logic in tuple/option/expected I would expect to be baked in here or otherwise into some trait impl on `T` perhaps. I don't think we'd want to duplicate things and have some using `PartialEq` and some using manual impls.

alexcrichton commented on crates/wasmtime/src/component/mod.rs:
```diff
@@ -9,13 +9,17 @@ mod instance;
 mod linker;
 mod matching;
 mod store;
+mod types;
```
GIven how many types are in this module and how `Type` isn't the only export I think it may actually make sense to have this be a `pub mod types` and don't reexport `Type` from inside. Otherwise external users can't name types like `Field` and such.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+};
+
+#[derive(Clone)]
+pub struct Handle<T> {
```
This is a bit bike-sheddy, but instead of 

```rust
pub struct Handle<T> { ... }
pub struct TupleIndex(TypeTupleIndex);
pub enum Type {
    // ...
    Tuple(Handle<TupleIndex>),
    // ...
}
```

could this instead be:

```rust
struct Handle<T> { ... }
pub struct Tuple(Handle<TypeTupleIndex>);
pub enum Type {
    // ...
    Tuple(Tuple),
    // ...
}
```

That I think will allow us to be a bit more flexible in the future with internal representations ideally.

alexcrichton replied:
I think this organization also helps method discovery since today there's lots of impls for `Handle<T>` for different `T` but for `Tuple` there will be just the methods related to tuples.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+        } else {
+            Err(anyhow!("cannot make list from {} type", self.desc()))
+        }
+    }
```
I think these methods may make more sense on `List` (in the values module) along the lines of:


```rust
impl List {
    pub fn new(elements: Box<[Val]>, ty: types::List) -> Result<List> {
        // ...
    }
}
```

(using the different structure of types I mentioned above).

Then there could be `impl From<List> for Val` to convert to a `Val` as necessary.

dicej replied:
That's what I originally considered, but I ended up using these methods on `Type` instead in the interest of ergonomics.  You can see them used in the dynamic.rs tests, along with `Type::nested`.  This avoids forcing the programmer to write a bunch of `if let Type::List(ty) = ty { List::new(elements, ty) } else { unreachable!() }`.

This is the ergonomic challenge I was referring to earlier in the discussion about whether `Val`s need to be associated with a specific `Type`.  I'm certainly willing to reconsider; I just want to understand how you expect this API to be used vs how I'm using it in the dynamic.rs tests.  Do you not expect real-world code to need a bunch of `if let ... { ... } else { unreachable!() }` when building `Val`s from `Type`s in the scheme you outlined?

alexcrichton replied:
Ah ok thanks for pointing that out! I think this may be a case where I'm used to slightly more verbose Rust-based idioms perhaps? For example the `Type::new_list` API as-is can return two kinds of errors (not a list type or element of the wrong type) where what I would otherwise expect is to split those two error cases across two functions. Instead of:

```rust
    let input = ty.new_list(Box::new([
        Val::U32(32343),
        Val::U32(79023439),
        Val::U32(2084037802),
    ]))?;
```

I'd expect:

```rust
    let input = ty.unwrap_list().new(Box::new([
        Val::U32(32343),
        Val::U32(79023439),
        Val::U32(2084037802),
    ]))?;
```

(or similar)

I don't mean to say we should go all the way to writing a bunch of wordy `if let` but I think helper methods can help scope precisely what errors are coming out of each function. In this case `unwrap_list` would be a clear indication of "I expect this type to be a list" and the `new` error would be clearly "I expect each element to be of this type".

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+    ///
+    /// For example, this will return the field types in order of declaration for a record type.  It will return
+    /// the element type for a list type and nothing for non-composite types like bool, u8, and string.
+    pub fn nested(&self) -> Box<[Type]> {
```
Is this still necessary? This seems like a bit of an odd query that I'm not sure what it would be used for

dicej replied:
Please see my comment above about ergonomics and how it is used in the dynamic.rs tests.

dicej replied:
@alexcrichton any further concerns about this method?  I'm going to mark this "resolved" for now, but feel free to unresolve it.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+                self.desc(),
+                value.ty().desc()
+            ))
+        }
```
This doesn't necessarily have to be handled here, but I think this typecheck will eventually want to be more recursive for better error mesages. Right now I think you could get `expected record, got record` which doesn't describe which record fields mismatch or similar.

dicej replied:
Any e.g. record field mismatch would have been caught already in `Type::new_record` before we got this far.  The only time you'd see `expected record, got record` would be if you used `Val`s from one component in another, or if your component had two identical `record`s (i.e. same field names and types) and you mixed up which one you created the `Val` for.  Either way, more recursion wouldn't necessarily help.  This is somewhat analogous to when `rustc` reports a type mismatch between `foo::Bar` and `foo::Bar` because you're using two different versions of the `foo` crate.  It's even worse here since types don't really have names in the component model.

But yes, I agree we should give a friendlier diagnostic here, e.g. hinting based on whether `Handle::index` or `Handle::types` are different, or both.

dicej replied:
I've improved the error message so you'll never get `expected record, got record`.  I attempted to go much further than that and recursively compare the types, but reusing `typed::typecheck_tuple` etc. turned out to be difficult in a dynamic context, and I didn't want to duplicate them either.

Also, I'm having trouble finding guidance in https://github.com/WebAssembly/component-model about when interface types are to be considered equal and when not, so I'm not sure what the correct logic is anyway.  Are types considered equal if they are structurally identical?

alexcrichton replied:
I don't think we'll get a ton of guidance from the component model itself, but for now from Wasmtime's perspective given how trampolines and lifting/lowering/etc are all implemented we require exact structural equality for the embedder API. I think the intention is that subtyping would eventually be used but we don't have a great means of implementing that right now.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+            Type::String | Type::List(_) => {
+                // These won't fit in func::MAX_STACK_RESULTS as of this writing, so presumably we should never
+                // reach here
+                unreachable!()
```
I think it would still be worth filling this in for other reasons though. For example all of the parameters to a host function are lifted which is where these could show up inline.

Eventually we'll want an equivalent to `crate::Linker::func_new` within `component::Linker` although it doesn't have to be added as part of that PR. When that's added this'll get exercised I believe.

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1025 @@
+        .collect()
+}
+
+fn lift_variant<'a>(
+    mut types: impl ExactSizeIterator<Item = Type>,
+    store: &StoreOpaque,
+    options: &Options,
+    src: &mut impl Iterator<Item = &'a ValRaw>,
+) -> Result<(u32, Val)> {
+    let len = types.len();
+    let discriminant = next(src).get_u32();
+    let ty = types
+        .nth(discriminant as usize)
+        .ok_or_else(|| anyhow!("discriminant {} out of range [0..{})", discriminant, len))?;
+    let value = ty.lift(store, options, src)?;
+    Ok((discriminant, value))
+}
```
I think that this is going to need special logic to skip extra fields within `src` to ensure that `flatten_count` entries are always popped off that iterator.

alexcrichton reviewed (commented):
Looking good! I left a bunch of initial comments and I'll take another pass when they're addressed.

dicej commented:
Thanks so much for the thorough review, @alexcrichton.  I've addressed some of the issues you raised (and marked the corresponding conversations "resolved") and will look at the others tomorrow.

alexcrichton reviewed (commented):
This is looking great, thanks again for all this! I think with some minor fixes here and there this is basically good to go.

Personally I feel that `Type::nested` should still get removed. I think that the usage in tests can be `.unwrap_$thing().field_accessor()` perhaps?

dicej commented on crates/wasmtime/src/component/func/typed.rs:
```diff
@@ -730,6 +510,50 @@ forward_impls! {
+
+macro_rules! forward_list_lifts {
+    ($($a:ty,)*) => ($(
+        unsafe impl <T: Lift + Lower> Lift for $a {
```
This is what I get when I remove the `+ Lower`:

```
error[E0277]: the trait bound `T: component::func::typed::Lower` is not satisfied
    --> crates/wasmtime/src/component/func/typed.rs:536:31
     |
536  |           unsafe impl <T: Lift> Lift for $a {
     |                                 ^^^^ the trait `component::func::typed::Lower` is not implemented for `T`
...
550  | / forward_list_lifts! {
551  | |     Box<[T]>,
552  | |     std::rc::Rc<[T]>,
553  | |     std::sync::Arc<[T]>,
554  | |     Vec<T>,
555  | | }
     | |_- in this macro invocation
     |
note: required because of the requirements on the impl of `component::func::typed::Lower` for `[T]`
    --> crates/wasmtime/src/component/func/typed.rs:1006:16
     |
1006 | unsafe impl<T> Lower for [T]
     |                ^^^^^     ^^^
note: required because of the requirements on the impl of `component::func::typed::ComponentType` for `Box<[T]>`
    --> crates/wasmtime/src/component/func/typed.rs:475:37
     |
475  |           unsafe impl <$($generics)*> ComponentType for $a {
     |                                       ^^^^^^^^^^^^^
...
504  | / forward_impls! {
505  | |     (T: Lower + ?Sized) &'_ T => T,
506  | |     (T: Lower + ?Sized) Box<T> => T,
     | |                         ^^^^^^
507  | |     (T: Lower + ?Sized) std::rc::Rc<T> => T,
...    |
510  | |     (T: Lower) Vec<T> => [T],
511  | | }
     | |_- in this macro invocation
note: required by a bound in `component::func::typed::Lift`
    --> crates/wasmtime/src/component/func/typed.rs:444:32
     |
444  | pub unsafe trait Lift: Sized + ComponentType {
     |                                ^^^^^^^^^^^^^ required by this bound in `component::func::typed::Lift`
     = note: this error originates in the macro `forward_list_lifts` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `T`
     |
536  |         unsafe impl <T: component::func::typed::Lower Lift> Lift for $a {
     |                         +++++++++++++++++++++++++++++
...
```

alexcrichton replied:
Oh that bug is [here](https://github.com/bytecodealliance/wasmtime/blob/d89c262657e98caf0195c07ac3fd72126e1f8ec3/crates/wasmtime/src/component/func/typed.rs#L695) where right now there's `<T: Lower> ComponentType for Box<T>` where that should be `T: ComponentType`. I was trying to be "too clever" by shoving it all into one macro but I think it can be split by refactoring the forward macro into multiple macros

dicej commented:
@alexcrichton I believe I've addressed everything you mentioned.  I'm still unsure about how `Type` equality should work, but I'm hoping the current implementation is good enough to start with and can be refined later.

alexcrichton commented on crates/wasmtime/src/component/values.rs:
```diff
@@ -0,0 +1,582 @@
+
+impl Option {
+    /// Instantiate the specified type with the specified `value`.
+    pub fn try_new(ty: &types::Option, value: std::option::Option<Val>) -> Result<Self> {
```
Idiomatically I think it's fine to call these constructors `new` rather than `try_new` since `try_*` is typically only used if the base name is already a method elsewhere (and otherwise there's no `Option::new`)

alexcrichton commented on crates/wasmtime/src/component/types.rs:
```diff
@@ -0,0 +1,1023 @@
+    }
+
+    /// Deserialize a value of this type from core Wasm stack values.
+    pub(crate) fn lift<'a>(
```
Would it be possible to move `lift`/`load` to the `Val` module and keep all the internals of all the `Val` private (e.g. not `pub(crate)`)?

alexcrichton reviewed (approved):
Looks great to me, thanks again for your work on this! I've got some follow-ups to your comments below and a minor nit or two, but otherwise this looks good-to-go.

dicej commented:
@alexcrichton I've addressed your latest feedback and opened https://github.com/bytecodealliance/wasmtime/issues/4522.

alexcrichton commented:
Thanks again for this @dicej!
